### ISSUE TITLE
[roadmap] Minetest 0.5.0 objectives

### ISSUE THREAD

**Issue by nerzhul** [minetest/minetest#2370]
This issue permits to have a view on minetest objectives for the 0.5.0 release, which will happen in the future (4-6 months ? later ? we don't know).

If you have ideas for the next release, please add it there
# Objectives
## Breaking changes
- [ ] Network protocol fixes (in progress)
- [ ] Network layer change (?)
- [x] Reworked client authentication (#1943)
- [ ] Custom player breath (#2246)
- [ ] Custom player health (#2246)
- [x] Server side breath
- [ ] Server side damages (#1894)
- [x] Generic stats
## New features
- [ ] Reworked modstore (design & categories (#2196)
- [ ] UTF-8 support
- [x] Colored chat (#2136)
- [ ] Meta set nodedef (#1118)
- [ ] Custom cloud height
## New Lua API features
- [x] `on_punch` callback to detect who punch who
- [x] Mod security
- [x] Client-side Lua?
## Improvements
- [ ] Inprove inventory performance
- [ ] Re-design formspec (#1399)
- [ ] Support non-qwerty keyboards
- [x] Enable locking for SQLite database
## Code organisation
- [ ] Organize code into a shallow directory structure (e.g., `gui*` to `client/`)
- [ ] Rename and organize files to follow naming conventions (should be `foo_bar` (`content_abm.cpp`), but `foobar` (`defaultsettings.cpp`), `fooBar` (`guiTable.cpp`), and `foo-bar` (`database-sqlite3.cpp`) variations exist)

**Comment by C1ffisme** [comment]
Did you seriously add the Roadmap label just for this ONE issue?

Add Voxel Area Entities to the New features list. Trust me, it will give us a feature minecraft does not have yet, and will let many more players see what open-source is all about.

**Comment by SmallJoker** [comment]
@C1ffisme, would you expect more roadmaps for this milestone?

Allow us to rotate objects in all directions and add "undocked" nodes, so it's possible push those away like real items.

**Comment by RobertZenz** [comment]
@nerzhul There's #1399 around, which is most likely a very big feature/work and is marked for 0.5.

@C1ffisme But FLOSS is not about "we have so many more features than everyone else", it's about "you can change it and do with it what you want".

**Comment by RHRhino** [comment]
Some ideas:
* Wielded light (carrying torches in your hand should light up caves)
* Colored light
* Minetest-to-Blender export (I want to render my buildings/worlds ^.^ )
* Randomized textures

@SmallJoker "undocked nodes" :+1: How about pushing falling nodes?
@C1ffisme "Voxel Area Entities" :+1:

**Comment by SmallJoker** [comment]
@RHRhino, pushing falling nodes which can be attached to other falling nodes

**Comment by JakubVanek** [comment]
Some of my suggestions:
- This is relatively big change: #2196
- https://github.com/minetest/minetest_game/issues/408

**Comment by Amaz1** [comment]
I think #2080 would be a good addition.

**Comment by JakubVanek** [comment]
Lua API: #2250, fixes

**Comment by Novatux** [comment]
Lua API: #1118, and VAEs would be two useful changes, perhaps mod security as a precursor to client-side Lua (which would probably be for later, 0.5.0 seems too soon)

**Comment by MirceaKitsune** [comment]
Those are some really good points so far! I'm glad more attention is given to the networking system, which I think is still highly unoptimized and slow. Not to mention the lack of prediction for meta inventories, and how laggy and ugly it is to move items in a chest on most servers! Please read my threads on both issues, in case anyone is interested:

https://forum.minetest.net/viewtopic.php?f=7&t=10697&p=163860

https://forum.minetest.net/viewtopic.php?f=6&t=9527&p=145064

Now if only yaw interpolation was also on this list :)

**Comment by C1ffisme** [comment]
@RobertZenz True, I open my mouth and not everything I say/type is what I wanted to say. I talk faster than I can think of words to express myself. I really meant something more like, newbies should learn that there is a reason to play minetest over minecraft.

**Comment by 4Evergreen4** [comment]
I also have a few suggestions:
* #1176
* An entity/player.on_puch callback

An on_pucnh function would be useful for detecting who killed who or who killed what.

**Comment by hunter688** [comment]
Vote for #1072 #2338 #2339

**Comment by C1ffisme** [comment]
Are we going to add VAEs, or no? I didn't see them added to the list.

Another Question : Does this roadmap include minetest_game changes or just changes to the minetest engine? (Not that I want VAEs developed in minetest_game, I still want them in the engine.)

**Comment by MT-Modder** [comment]
@C1ffisme I believe the decision was 0.5 was too soon for VAE's.

**Comment by C1ffisme** [comment]
@MT-Modder Ah, okay. Maybe 0.6, or 1.0.

I sure do hope there are some big gameplay changes, though.

Should a roadmap be added to minetest_game?

**Comment by MT-Modder** [comment]
One for minetest_game will probably be added. It looks like most changes are major bug fixes and a few new modding api features.

Here is the current list: https://github.com/minetest/minetest/milestones/0.5.
More goals will probably be added and some might be pushed toward the next milestone.

**Comment by paramat** [comment]
Celeron55 has posted a roadmap in the forum https://forum.minetest.net/viewtopic.php?f=7&t=9177&hilit=roadmap

**Comment by Wuzzy2** [comment]
I vote for a few modding-related issues:
* #1176 (+accelleration and setter functions)
* #1496
* #2246 (a couple of modders want this, so I don't understand why this is marked “Low Priority”)

**Comment by nerzhul** [comment]
@paramat yes, it's a general roadmap. I think, because @ShadowNinja is working on database rework we should include:
- Enable locking so that the database cannot be opened by multiple instances of Minetest
- Resolve issues or close it.

**Comment by MT-Modder** [comment]
I would like to see animated inventory images added sometime in the future. That would be cool :-)

**Comment by ShadowNinja** [comment]
@nerzhul: All of our DB backends already implement cross-process locking (except `dummy` of course).  They also allow you to run other things like mappers while the server's running, which is good because it doesn't require copying or downtime.

**Comment by nerzhul** [comment]
@ShadowNinja then maybe we need to update celeron55 roadmap then, because it's always set

**Comment by figec** [comment]
>  UTF-8 support

This is especially important. This will fix the most serious lack of the game.

**Comment by Holyphoenix** [comment]
+1 wielding torches.

**Comment by SmallJoker** [comment]
@Holyphoenix The idea is not new and it's still not good until there is dynamic lightening and light emitting support for entities.

**Comment by Holyphoenix** [comment]
@SmallJoker - https://github.com/minetest/minetest/pull/816#issuecomment-78314198

Anyway, it would be a lot better once dynamic lighting and light emitting support for entities are added, especially since other people will be able to see the light from your torch.  So +1 dynamic lighting and +1 light emitting entities!

**Comment by CasimirKaPazi** [comment]
What is with case-insensitive player names? #1511

**Comment by proller** [comment]
https://github.com/minetest/minetest/compare/master...sapier:generic_player_stats

**Comment by hunter688** [comment]
Network layer change mean change from UDP to TCP?

**Comment by neoascetic** [comment]
I guess mapgen v7 should be finished before 0.5 release ;)

**Comment by MirceaKitsune** [comment]
@neoascetic Someone brave enough needs to poke hmmmm with a stick about that :) Anyway I perfectly agree... one of the things I am most interested in seeing is MGV7 being finished at last!

**Comment by MirceaKitsune** [comment]
Oh... there is one more thing which I believe should be added to the list, and IMO must be fixed before the 0.5 release: Pathfinding. minetest.find_path is basically unusable, because the command almost always fails to generate a path and is extremely buggy. My old report can be found here:

https://github.com/minetest/minetest/issues/1701

**Comment by neoascetic** [comment]
> Custom cloud height

@nerzhul AFAIK it is already supported with `cloud_height` option. But IMO it should be improved: cloud height should be set on server-side and not client side (because on server `water_level` may be set to value higher than client's `cloud_height` and user will get clouds under the water). What do you think on this?

**Comment by nerzhul** [comment]
Checked some done features

**Comment by paramat** [comment]
> I guess mapgen v7 should be finished before 0.5 release ;)

> Someone brave enough needs to poke hmmmm with a stick about that :) Anyway I perfectly agree... one of the things I am most interested in seeing is MGV7 being finished at last!

Hehe that's actually my responsibility, hmmmm was fed up working on it so i offered to improve it and maintain it, i have worked on it a huge amount, he likes the result and i consider it the best mapgen. All that's left is improving the terrain noise parameters and adding the biome system, we have fixed the last remaining bugs and issues so just a few weeks more now. If not 0.4.13 then certainly for the big release. Mgv5 will be stable with biomes too.

**Comment by ProgrammingChicken** [comment]
Yess, been waiting for custom health and breath.
When 5.0's out, I'll finally be able to finish my mod.

**Comment by RobertZenz** [comment]
I know I'm kinda late with that, but can we get #2151 on the list? It is a real pain if you're trying to write your own mapgen and want to spawn trees and it takes a half minute upwards for all trees to be spawned. I'm very sure that such mods like moretrees would also benefit greatly from it.

**Comment by neoascetic** [comment]
+1 for wield light (#816)

**Comment by wilkgr76** [comment]
I'd really like:
- light wielding
- custom skins (I know there are plenty of skin mods, but how about integrating it like voxelands?)
- client-side lua (to prevent griefing, unknown blocks, etc, prevent any mods in the  "clientmods" folder from running if they define any new nodes, tools, etc.)
- quicker connection to multiplayer (Use MD5, maybe, so that the server doesn't send unnecessary files)

**Comment by ShadowNinja** [comment]
@RHRhino: MT-to-Blender is something to be done with an external script.

@wilkgr76
custom skins: That's a game issue.
quick connection: A media cache is already implemented, although it uses SHA-1, not MD5.

**Comment by neoascetic** [comment]
Client side lua request: #3440

**Comment by asl97** [comment]
@neoascetic (iirc) client side lua has long been listed,  there is also two page on the dev wiki

http://dev.minetest.net/TODO#Big.2C_long_term_goals_.28Client.29
http://dev.minetest.net/Client_scripting_plans

**Comment by paramat** [comment]
0.5.0 is now 0.4.14, can we remove the '0.5' milestone? If so i guess those issues/PRs should be set with milestone 0.4.14?

**Comment by neoascetic** [comment]
We will wait for 0.4.14 for two more years then, because amount of changes is very big.

**Comment by 0-afflatus** [comment]
￼ > Custom player breath (#2246)
￼ > Custom player health (#2246)

These should be part of a generic, extensible stat system which would allow hunger, suffocation and other arbitrary stats to be easily soft-coded.

**Comment by wilkgr76** [comment]
@0-afflatus Yes. +1

**Comment by iqualfragile** [comment]
https://github.com/minetest/minetest/issues/4184
is concerned with networking and would most likely be a backward compatibility breaking change

**Comment by Wuzzy2** [comment]
Can you add “Use SemVer” on the roadmap?
This was discussed here:
https://forum.minetest.net/viewtopic.php?f=3&t=10848&hilit=semver#p166143

It was more or less accepted by the developers as far I can remember.

**Comment by Wuzzy2** [comment]
Can you tick the “custom cloud height” from the TODO list? It is already possible for a long time with the setting `cloud_height` in `minetest.conf`. But this setting is client-side only. Maybe you want to clarify what you wanted to add.

**Comment by bigfoot547** [comment]
Player fov #1075.

**Comment by nerzhul** [comment]
i close this issue as we are focused on CSM and bugfixes.

**Comment by thePalindrome** [comment]
I'd say we should leave this open seeing as we want to incorporate as many features as possible into this release because we are breaking compatibility. Bug fixes are typically compatible with earlier versions, not to mention that 0.5 might not even be the next version!

I vote we keep this open to continue discussion on these features and to provide a place for interested parties to check on progress.

**Comment by nerzhul** [comment]
@thePalindrome it seems 0.5 is delayed for a long amount of time at this moment, except if we decide CSM and all master changes are sufficient to justify a major version bump (and i think it should)

**Comment by nerzhul** [comment]
This issue is deprecated and can be replaced by another and https://github.com/minetest/minetest/projects/4

**Comment by Wuzzy2** [comment]
Does the deprecation also mean that all unfinished things of the TODO list in the first post have been discarded?
I haven't seen seperate issues covering the missing items on the TODO list.

Please don't forget the things on the TODO list (unless you discarded it completely).

**Comment by nerzhul** [comment]
s/discarded/delayed for re-discussion/g

**Comment by Fixer-007** [comment]
Built in way to show text in game (for signs and other).

**Comment by NewbProgrammer101** [comment]
And also fix the water placement bugs in the floatlands of mapgen v7.
